scipy/special/_logsumexp.py

```diff
@@ -212,4 +212,65 @@ def softmax(x, axis=None):
     """
 
     # compute in log space for numerical stability
-    return np.exp(x - logsumexp(x, axis=axis, keepdims=True))
+    # subtraction can cause cancellation when x and logsumexp(x) are close to each other.
+    # return np.exp(x - logsumexp(x, axis=axis, keepdims=True))
+    return np.exp(log_softmax(x, axis=axis))
+
+
+def log_softmax(x, axis=None):
+    r"""
+    Logarithm of softmax function::
+
+        log_softmax(x) = log(softmax(x))
+
+    Parameters
+    ----------
+    x : array_like
+        Input array.
+    axis : int or tuple of ints, optional
+        Axis to compute values along. Default is None and softmax will be
+        computed over the entire array `x`.
+
+    Returns
+    -------
+    s : ndarray
+        An array the same shape as `x`. Exponential of the result will sum to 1 along the
+        specified axis.
+
+    Notes
+    -----
+    `log_softmax` is more accurate than ``np.log(softmax(x))`` with inputs that make 
+    `softmax` saturate (see examples below).
+    
+    .. versionadded:: 1.5.0
+
+    Examples
+    --------
+    >>> from scipy.special import log_softmax
+    >>> from scipy.special import softmax
+    >>> np.set_printoptions(precision=5)
+
+    >>> x = np.array([1000.0, 1.0])
+
+    >>> y = log_softmax(x)
+    >>> y
+    array([   0., -999.])
+
+    >>> y = np.log(softmax(x))
+    RuntimeWarning: divide by zero encountered in log
+    >>> y
+    array([  0., -inf])
+
+    """
+
+    x = _asarray_validated(x, check_finite=False)
+
+    x_max = np.amax(x, axis=axis, keepdims=True)
+
+    if x_max.ndim > 0:
+        x_max[~np.isfinite(x_max)] = 0
+    elif not np.isfinite(x_max):
+        x_max = 0
+
+    tmp = x - x_max
+    return tmp-logsumexp(tmp, axis=axis, keepdims=True)
```

Comment:
I think it would make more sense to not try and reuse the `logsumexp` logic here-we will end up looking for the max again when we already know it is 0.

Reply:
In the latest commit, I changed the function code so that it does not reuse logsumexp.